Initial Developer Documentation (#148)
10up/plugin-check: docs/creating-a-runtime-check.md

# Creating a Runtime Check

Runtime checks are used to perform tests by executing the plugin's code.

The Plugin Checker uses a number of interfaces to build out runtime checks.

The Plugin Checker provides the `Runtime_Check` interface, which is used to identify a runtime check. This interface does not contain any methods but serves as a marker for runtime checks.

```php
use WordPress\Plugin_Check\Checker\Check_Result;
use WordPress\Plugin_Check\Checker\Runtime_Check;
use WordPress\Plugin_Check\Traits\Stable_Check;

class My_Custom_Check implements Runtime_Check {
  
  use Stable_Check;
  
  public function get_categories() {
    // Return an array of check categories.
    // See the `WordPress\Plugin_Check\Checker\Check_Categories` class for available categories.
  }
  
  public function run( Check_Result $result );
    // Handle running the check and adding warnings or errors to the result.
  }
}
```

> In addtion, there is also the `Preparation` interface, which can be implemented by a check to prepare the WordPress environment. This interface defines a `prepare()` method which is used to run the logic required to prepare the environment before the check before is run.

Reviewer: ```suggestion
In addition, there is also the `Preparation` interface, which can be implemented by a check to prepare the WordPress environment. This interface defines a `prepare()` method which is used to run the logic required to prepare the environment before the check before is run.
```

```php
use WordPress\Plugin_Check\Checker\Checks\Abstract_Runtime_Check;

class My_Custom_Check extends Abstract_Runtime_Check {
  /**
   * Check Preparation.
   */
  public function prepare() {
    // Create a test post ahead of running the check.
    $post_id = wp_insert_post(
      array(
        'post_title'   => 'Test post',
        'post_content' => 'This is a test post.',
        'post_type'    => 'post',
      )
    );

    // Return the clean up function.
    return function() use ( $post_id ) {
      // Remove the test post created.
      wp_delete_post( $post_id );
    }
  }
}
```

### Shared Preparations

Shared preparations are individual classes created to handle specific preparation logic in a single place.

The [`Demo_Posts_Creation_Preparation`](https://github.com/10up/plugin-check/blob/trunk/includes/Checker/Preparations/Demo_Posts_Creation_Preparation.php) in the Plugin Checker is an example of a preparation class.

Preparation classes are useful for implementing the same preparation logic across multiple checks using the concept of shared preparations.

Shared preparations are used to prevent running the same preparations multiple times.

> Before running checks against a plugin all shared preparations are collected and processed to remove any shared preparations that are the same. 

Reviewer: ```suggestion
Before running checks against a plugin all shared preparations are collected and processed to avoid duplicate execution of any shared preparations that are the same. 
```

Any preparation class can be used as a shared preparation including [preparations already available in the Plugin Checker](https://github.com/10up/plugin-check/blob/trunk/includes/Checker/Preparations).

Check classes can define the shared preparations the use by using the `With_Shared_Preparation` interface.

The Check class should then implement the `get_shared_preparations()` method defined by the interface. This method returns an map of shared preparations where the preparation class name is the key and an array of constructor parameters as the value.

> Below is an example of how the `Enqueued_Scripts_Size_Check` uses shared preparations.

Reviewer: ```suggestion
Below is an example of how the ` Demo_Posts_Creation_Preparation` preparation class can be used to generate a demo post for every viewable post type, which are then removed again after running the checks.
```

```php
/**
 * Returns an array of shared preparations for the check.
 *
 * @return array Returns a map of $class_name => $constructor_args pairs. If the class does not
 *               need any constructor arguments, it would just be an empty array.
 */
public function get_shared_preparations() {
  $demo_posts = array_map(
    function( $post_type ) {
      return array(
        'post_title'   => "Demo {$post_type} post",
        'post_content' => 'Test content',
        'post_type'    => $post_type,
        'post_status'  => 'publish',
      );
    },

>     $this->viewable_post_types

Reviewer: This variable should be replaced with the logic to get the viewable post types, since the class property is not part of this example code.

# Creating a Runtime Check

> Runtime checks are used to perform tests by executing the plugin's code.

Reviewer 2: ```suggestion
Runtime checks involve executing the plugin's code to perform tests.
```

More active and concise.

> The Plugin Checker uses a number of interfaces to build out runtime checks.

Reviewer 2: ```suggestion
The Plugin Checker utilizes various interfaces to construct runtime checks.
```
More active & dynamic.

> The Plugin Checker provides the `Runtime_Check` interface, which is used to identify a runtime check. This interface does not contain any methods but serves as a marker for runtime checks.

Reviewer 2: ```suggestion
The Plugin Checker provides the `Runtime_Check` interface, designed to identify a runtime check. While this interface does not contain any methods, it serves as a marker for runtime checks.
```

Better clarification of what the `Runtime_Check` interface is design to do.

```php
use WordPress\Plugin_Check\Checker\Check_Result;
use WordPress\Plugin_Check\Checker\Runtime_Check;
use WordPress\Plugin_Check\Traits\Stable_Check;

class My_Custom_Check implements Runtime_Check {

  use Stable_Check;

  public function get_categories() {
    // Return an array of check categories.
    // See the `WordPress\Plugin_Check\Checker\Check_Categories` class for available categories.
  }

  public function run( Check_Result $result ) {
    // Handle running the check and adding warnings or errors to the result.
  }
}
```

> In addition, there is also the `Preparation` interface, which can be implemented by a check to prepare the WordPress environment. This interface defines a `prepare()` method which is used to run the logic required to prepare the environment before the check before is run.

Reviewer 2: ```suggestion
In addition, the Plugin Checker includes the `Preparation` interface, allowing a check to prepare the WordPress environment. This interface defines a `prepare()` method used to execute the necessary logic before the check is run.
```

We can be less specific about the fact that it prepares an environment, and simply state that it runs the necessary logic before a check is executed, as we already state that it has to do with environment preparation.

> Both these interfaces are implemented in the `Abstract_Runtime_Check` class, which developers can use when building out their own runtime checks. This class defines both the `prepare()` and `run()` methods which are required to be implemented for every runtime check.

Reviewer 2: ```suggestion
Both of these interfaces, `Runtime_Check` and `Preparation`, are implemented within the `Abstract_Runtime_Check` class, providing developers with a foundation for building their own runtime checks. This class defines the `prepare()` and `run()` methods, which are required to be implemented for every runtime check.
```

Better flow and readability, and also explicitly states which interfaces we are referring to for better qualification.

> Below is the basic scaffold when creating a custom runtime check.

Reviewer 2: ```suggestion
Below is the basic framework for creating a custom runtime check.
```

it's not a really a scaffold, there is no command or means to _scaffold_ this code, it is merely a framework for configuring a custom runtime check. Not a total blocker for me though.

```php
use WordPress\Plugin_Check\Checker\Check_Result;
use WordPress\Plugin_Check\Checker\Checks\Abstract_Runtime_Check;
use WordPress\Plugin_Check\Traits\Stable_Check;

class My_Custom_Check extends Abstract_Runtime_Check {

  use Stable_Check;
  
  public function get_categories() {
    // Return an array of check categories.
    // See the `WordPress\Plugin_Check\Checker\Check_Categories` class for available categories.
  }

  public function prepare() {
    // Handle the Checks preparations and return a cleanup function.
  }

  public function run( Check_Result $result ) {
    // Handle running the check and adding warnings or errors to the result.
  }
}
```

## Preparations

> Preparations in the Plugin Checker are used to set up the environment before running checks, involving logic like adding actions and filters, and creating necessary test content. They utilize the `Preparation` interface, requiring the implementation of a `prepare()` method, and return cleanup functions to revert changes made during preparation.

Reviewer 2: ```suggestion
In the Plugin Checker, preparations are employed to set up the environment before executing checks. These preparations involve implementing logic such as adding actions and filters, and creating the necessary test content. They utilize the `Preparation` interface, which mandates the implementation of a `prepare()` method and requires returning cleanup functions to revert changes made during preparation.
```

More concise, better flow & clarity.

> Three different approaches can be used for creating preparations based on the circumstances:

Reviewer 2: ```suggestion

Based on the circumstances, preparations can be created using three different approaches:
```

Improved clarity, more active. Adds missing line prior to sentence.

* Check preparations
* Shared preparations
* Global preparations

### Check preparations

> Check preparations are created using the `prepare()` method within a check class.

Reviewer 2: ```suggestion
Check preparations are created by defining the `prepare()` method within a check class.
```

More concise representation of the process.

> This type of preparation is useful for carrying out the logic to prepare the environment that is specific to the check that implements it.

Reviewer 2: ```suggestion
This type of preparation is valuable for executing the logic necessary to prepare the environment specifically tailored to the check that implements it.
```

Emphasizes the importance and significance of the preparation type. More concise and coherent.

> Below is an example of a preparation that creates a test post and returns a cleanup function that deletes that post after the check is run.

Reviewer 2: ```suggestion
Below is an example of a preparation that generates a test post and includes a cleanup function - which is returned as part of the `prepare()` method - to delete the post after the check is executed.
```

```php
use WordPress\Plugin_Check\Checker\Checks\Abstract_Runtime_Check;

class My_Custom_Check extends Abstract_Runtime_Check {

  /**
   * Check Preparation.
   */
  public function prepare() {
    // Create a test post ahead of running the check.
    $post_id = wp_insert_post(
      array(
        'post_title'   => 'Test post',
        'post_content' => 'This is a test post.',
        'post_type'    => 'post',
      )
    );

    // Return the clean up function.
    return function() use ( $post_id ) {
      // Remove the test post created.
      wp_delete_post( $post_id );
    }
  }
}
```

### Shared Preparations

> Shared preparations are individual classes created to handle specific preparation logic in a single place.

Reviewer 2: ```suggestion
Shared preparations are individual classes that consolidate specific preparation logic in a single, centralized location.
```

More concise & active.

The [`Demo_Posts_Creation_Preparation`](https://github.com/10up/plugin-check/blob/trunk/includes/Checker/Preparations/Demo_Posts_Creation_Preparation.php) in the Plugin Checker is an example of a preparation class.

> Preparation classes are useful for implementing the same preparation logic across multiple checks using the concept of shared preparations.

Reviewer 2: ```suggestion
Preparation classes are beneficial for implementing the same preparation logic across multiple checks, leveraging the concept of shared preparations.
```

Better flow and readability, using "beneficial" over "useful" to provide a stronger emphasis on the advantages of using preparation classes.

> Shared preparations are used to prevent running the same preparations multiple times.

Reviewer 2: ```suggestion
Shared preparations are used to avoid redundant execution of the same preparations.
```

> Before running checks against a plugin all shared preparations are collected and processed to avoid duplicate execution of any shared preparations that are the same. 

Reviewer 2: ```suggestion
Before executing checks against a plugin, all shared preparations are gathered and processed to prevent redundant execution of identical shared preparations.
```

Better flow and clarity.

Any preparation class can be used as a shared preparation including [preparations already available in the Plugin Checker](https://github.com/10up/plugin-check/blob/trunk/includes/Checker/Preparations).

> Check classes can define the shared preparations the use by using the `With_Shared_Preparation` interface.

Reviewer 2: ```suggestion
Check classes can define and utilize shared preparations by implementing the `With_Shared_Preparation`interface.
```

> The Check class should then implement the `get_shared_preparations()` method defined by the interface. This method returns an map of shared preparations where the preparation class name is the key and an array of constructor parameters as the value.

Reviewer 2: ```suggestion
The Check class should then implement the `get_shared_preparations()` method as defined by the interface. This method returns a map of shared preparations, with the preparation class name as the key and an array of constructor parameters as the value.
```

"As defined by the interface" has been added to provide better context on where the `get_shared_preparations()` method is derived from. 

Slightly better flow.

> Below is an example of how the `Demo_Posts_Creation_Preparation` preparation class can be used to generate a demo post for every viewable post type, which are then removed again after running the checks.

Reviewer 2: ```suggestion
Below is an example of how the `Demo_Posts_Creation_Preparation` preparation class can be used to generate demo posts for every viewable post type, which are then removed after the checks are executed.
```

Slight simplification of the closing parts of the sentence.

```php
/**
 * Returns an array of shared preparations for the check.
 *
 * @return array Returns a map of $class_name => $constructor_args pairs. If the class does not
 *               need any constructor arguments, it would just be an empty array.
 */
public function get_shared_preparations() {
  $demo_posts = array_map(
    function( $post_type ) {
      return array(
        'post_title'   => "Demo {$post_type} post",
        'post_content' => 'Test content',
        'post_type'    => $post_type,
        'post_status'  => 'publish',
      );
    },
    array_filter( get_post_types(), 'is_post_type_viewable' )
  );

  return array(
    Demo_Posts_Creation_Preparation::class => array( $demo_posts ),
  );
}
```

> A check can have both a check preparation and a shared preparations combined. When a check implements both types of checks it is worth noting that shared preparations are run before check preparations.

Reviewer 2: ```suggestion
A check can combine both a check preparation and shared preparations. It's worth noting that when a check implements both types, shared preparations are executed before check preparations.
```

Restructured for simplicity and clarity.

## Global preparations

> Plugin Checker also includes some preparation classes that are used to prepare the overall environment, i.e. before any runtime check. Global preparations cannot be controlled outside of the Plugin Checker at this point.

Reviewer 2: ```suggestion
Plugin Checker also includes some preparation classes designed to prepare the overall environment, i.e., before any runtime check. At this point, global preparations cannot be controlled outside of the Plugin Checker.
```

Restructured for better clarity and readability.

> An example for a global preparation is the [`Force_Single_Plugin_Preparation`](https://github.com/10up/plugin-check/blob/trunk/includes/Checker/Preparations/Force_Single_Plugin_Preparation.php) class, which ensures that only the plugin to check is actually active.

Reviewer 2: ```suggestion
An example of a global preparation is the [`Force_Single_Plugin_Preparation`](https://github.com/10up/plugin-check/blob/trunk/includes/Checker/Preparations/Force_Single_Plugin_Preparation.php) class, which ensures that only the plugin being checked is active.
```

`of` instead of `for` to be more grammatically correct. Improved clarity & simplicity in the closing parts.

## Amending the check result object

> Amending results to the `Check_Result` object works in the same way as static checks. [See the documentation here](./creating-a-static-check.md#amending-the-check-result-object) for details on adding results.

Reviewer 2: ```suggestion
Amending results within the `Check_Result` object functions similarly to static checks. For more information on adding results, [refer to the documentation available here](./creating-a-static-check.md#amending-the-check-result-object).
```

I'm unsure if this should read "Amending results within the `Check_Result` object" or "Amending results of the `Check_Result` object" @mukeshpanchal27 as I was unsure if this was an error?

Reviewer 3: The changes make sense to me. Thanks

# Creating a Runtime Check

> Runtime checks are used to perform tests by executing the plugins code.

Reviewer: ```suggestion
Runtime checks are used to perform tests by executing the plugin's code.
```

The Plugin Checker uses a number of interfaces to build out runtime checks.

> One interface is the `Runtime_Check` interface. This interface contains no methods, but is used to help identify a runtime check from static checks.

Reviewer: Related to my feedback on the static check docs, maybe clarify that this is the only _requirement_ for a runtime check to implement, as well as accompany with a very quick code example like `Custom_Check implements Runtime_Check {...`

In addtion, there is also the `Preparation` interface, which is used to define checks that have preparations. This interface defines a `prepare()` method which is used to run the logic required to prepare the environment before the check before is run.

Both these interfaces are implemented in the `Abstract_Runtime_Check` class, which developers can use when building out their own runtime checks. This class defines both the `prepare()` and `run()` methods which are required to be implemented for every runtime check.

Below is the basic scaffold when creating a custom runtime check.

```php

> use WordPress\Plugin_Check\Checker\Checks\Abstract_Runtime_Check;

Reviewer: Nit-pick: Since `Check_Result` is referenced below, it also needs to be imported here via `use`.

class Custom_Check extends Abstract_Runtime_Check {
  public function prepare() {
    // Handle the Checks preparations and return a cleanup function.
  }

>   public function run( Check_Results $results ) {

Reviewer: There are some references to "results" throughout, which need to be corrected :)
```suggestion
  public function run( Check_Result $result ) {
```

>     // Handle running the check and adding warnings or errors to the results.

Reviewer: ```suggestion
    // Handle running the check and adding warnings or errors to the result.
```

  }
}
```

## Preparations

Preparations are used to prepare the environment before running the check. Preparations can include any logic from activating specific themes or plugins, or creating test content required to perform the check.

Preparations will use the `Preparation` interface which defines a single `prepare()` method. This method is required to be implemented in runtime checks and handles any logic needed to prepare the environent.

All preparations return a clean up function that reverts the changes made by the method.

> When creating preparations there are 3 approachs that can be used depending on circumstance.

Reviewer: You explain a lot of this already in the technical overview. At a minimum, we should align the two sections to basically have the same content, or alternatively we shorten the one in the technical overview to just the most important thing, like 1-2 sentences.

```php
use WordPress\Plugin_Check\Checker\Checks\Abstract_Runtime_Check;

class Custom_Check extends Abstract_Runtime_Check {
  /**
   * Check Preparation.
   */
  public function prepare() {
    // Create a test post ahead of running the check.
    $post_id = wp_insert_post(
      array(
        'post_title'   => 'Test post',
        'post_content' => 'This is a test post.',
        'post_type'    => 'post',
      )
    );

    // Return the clean up function.
    return function() use ( $post_id ) {
      // Remove the test post created.
      wp_delete_post( $post_id );
    }
  }
}
```

### Preparation Classes

Preparation classes are individual classes created to handle specific preparation logic in a single place.

The [`Force_Single_Plugin_Preparation`](https://github.com/10up/plugin-check/blob/trunk/includes/Checker/Preparations/Force_Single_Plugin_Preparation.php) in the Plugin Checker is an example of a preparation class.

Preparation classes are useful for implementing the same preparation logic across multiple checks using the concept of shared preparations.

> ### Shared Preparations

Reviewer: These aren't really 2 separate ways: Preparation classes are needed to use the concept of shared preparations, so I think these sections need to be combined. Shared preparations require separation classes - you may not always need to implement a new preparation if what you want to do can be achieved with an existing preparation class, but conceptually they belong together.

So overall, there are only two ways to deal with preparations, either a check implements `Preparation` and comes with its own `prepare()` method, or a check implements `With_Shared_Preparations` and uses shared preparations, in which case writing a `Preparation` class may be needed.

Worth noting that both can also be combined, a check could have shared preparations plus a check-specific preparation.

Check classes can define the shared preparations the use by using the `With_Shared_Preparation` interface.

The Check class should then implement the get_shared_preparations() method defined by the interface. This method returns an map of shared preparations where the preparation class name is the key and an array of constructor parameters as the value.

Below is an example of how the `Enqueued_Scripts_Size_Check` uses shared preparations.

```php
	/**
	 * Returns an array of shared preparations for the check.
	 *
	 * @return array Returns a map of $class_name => $constructor_args pairs. If the class does not
	 *               need any constructor arguments, it would just be an empty array.
	 */
	public function get_shared_preparations() {
		$demo_posts = array_map(
			function( $post_type ) {
				return array(
					'post_title'   => "Demo {$post_type} post",
					'post_content' => 'Test content',
					'post_type'    => $post_type,
					'post_status'  => 'publish',
				);
			},
			$this->viewable_post_types
		);

		return array(
			Demo_Posts_Creation_Preparation::class => array( $demo_posts ),
		);
	}
```

> ## Check run method

Reviewer: Much of this section is generally applicable, even to static checks, particularly adding to the `Check_Result` instance. Maybe we should move that to its own section like "Amending the check result object", which we can then reference in both articles? Also see my overarching review comment in that regard.

  - `$line (int)` - The line on which the message occurred. Default 0 (unknown line).
  - `$column (int)` - The column on which the message occurred. Default 0 (unknown column).

In addition to adding messages, the `Check_Result` instance also provides access to the `Plugin_Context`. The plugin context is useful for getting the plugins path and urls when performing checks.

Below is an example showing how to access the plugin context and add messages using the `Check_Results` instance.

```php
/**
	 * Runs the check on the plugin and amends results.
	 *
	 * @since n.e.x.t
	 *
	 * @param Check_Result $result The check results to amend and the plugin context.
	 */
	public function run( Check_Result $result ) {

    // Get the absolute file path for a specific file in the plugin.
    $plugin_file = $result->plugin()->path( 'plugin-file.php' );
    
    // Run logic to test the plugin for warnings/errors...

    // When an issue is found add a warning.
    $result->add_message(
      false, 
			'Warning message content.', 
			array(
				'code'   => 'warning_code',
				'file'   => $pluging_file,
				'line'   => 1,
				'column' => 10,
			)
    );
  }
```

> ## Add the Check to the Plugin Checker

Reviewer: See my feedback on the static check documentation for the similar section. We should make sure those sections are aligned, they can have the exact same content.

# Creating a Runtime Check

Runtime checks are used to perform tests by executing the plugin's code.

The Plugin Checker uses a number of interfaces to build out runtime checks.

The Plugin Checker provides the `Runtime_Check` interface, which is used to identify a runtime check. This interface does not contain any methods but serves as a marker for runtime checks.

```php
use WordPress\Plugin_Check\Checker\Check_Result;

> 

Reviewer 3: Do we need to use Stable_Check/Experimental_Check trail here?

Author: @mukeshpanchal27 I've added these to each of the code examples where I thought it was necessary

class Custom_Check extends Runtime_Check {
  public function run( Check_Result $result );
    // Handle running the check and adding warnings or errors to the result.

>   }

Reviewer 3: Needs to add `get_categories()` method.

Author: @mukeshpanchal27 I've updated the code examples with the `get_categories()` method where necessary

# Creating a Runtime Check

Runtime checks are used to perform tests by executing the plugin's code.

The Plugin Checker uses a number of interfaces to build out runtime checks.

The Plugin Checker provides the `Runtime_Check` interface, which is used to identify a runtime check. This interface does not contain any methods but serves as a marker for runtime checks.

```php
use WordPress\Plugin_Check\Checker\Check_Result;
use WordPress\Plugin_Check\Checker\Runtime_Check;
use WordPress\Plugin_Check\Traits\Stable_Check;

> class Custom_Check extends Runtime_Check {

Reviewer: See my feedback on the static check code example.
```suggestion
class My_Custom_Check implements Runtime_Check {
```

Author: Updated

  
  use Stable_Check;
  
  public function get_categories() {

>     // Return an array of check categories.

Reviewer: See my feedback on the static check code example, here it would be helpful to point to where the available categories are defined.

Author: Updated in both instances

  }
  
  public function run( Check_Result $result );
    // Handle running the check and adding warnings or errors to the result.
  }
}
```

> In addtion, there is also the `Preparation` interface, which is used to define checks that have preparations. This interface defines a `prepare()` method which is used to run the logic required to prepare the environment before the check before is run.

Reviewer: ```suggestion
In addtion, there is also the `Preparation` interface, which can be implemented by a check to prepare the WordPress environment. This interface defines a `prepare()` method which is used to run the logic required to prepare the environment before the check before is run.
```

Author: Updated

Both these interfaces are implemented in the `Abstract_Runtime_Check` class, which developers can use when building out their own runtime checks. This class defines both the `prepare()` and `run()` methods which are required to be implemented for every runtime check.

Below is the basic scaffold when creating a custom runtime check.

```php
use WordPress\Plugin_Check\Checker\Check_Result;
use WordPress\Plugin_Check\Checker\Checks\Abstract_Runtime_Check;
use WordPress\Plugin_Check\Traits\Stable_Check;

class Custom_Check extends Abstract_Runtime_Check {

  use Stable_Check;
  
  public function get_categories() {
    // Return an array of check categories.
  }

  public function prepare() {
    // Handle the Checks preparations and return a cleanup function.
  }

  public function run( Check_Result $result ) {
    // Handle running the check and adding warnings or errors to the result.
  }
}
```

## Preparations

> Preparations in the Plugin Checker are used to set up the environment before running checks, involving logic like activating themes or plugins and creating necessary test content. They utilize the `Preparation` interface, requiring the implementation of a `prepare()` method, and return cleanup functions to revert changes made during preparation. Three different approaches can be used for creating preparations based on the circumstances.

Reviewer: See my previous comment, we shouldn't focus on activating themes and plugins here as this is too specific.

Also, before diving into the specific types of preparations, we should list them to clarify and connect the sections. Also, preparation classes and shared preparations are really referring to the same thing, so that is just one, not two approaches. On the other hand, the article is currently missing the concept of global preparations.
```suggestion
Preparations in the Plugin Checker are used to set up the environment before running checks, involving logic like adding actions and filters, and creating necessary test content. They utilize the `Preparation` interface, requiring the implementation of a `prepare()` method, and return cleanup functions to revert changes made during preparation.

Three different approaches can be used for creating preparations based on the circumstances:
* Check preparations
* Shared preparations
* Global preparations
```

Author: Thanks @felixarntz updated and added the new Global preparations section

> ### Check Preperations

Reviewer: ```suggestion
### Check preparations
```

Author: Updated

Check preparations are created using the `prepare()` method within a check class.

This type of preparation is useful for carrying out the logic to prepare the environment that is specific to the check that implements it.

Below is an example of a preparation that creates a test post and returns a cleanup function that deletes that post after the check is run.

```php
use WordPress\Plugin_Check\Checker\Checks\Abstract_Runtime_Check;

class Custom_Check extends Abstract_Runtime_Check {
  /**
   * Check Preparation.
   */
  public function prepare() {
    // Create a test post ahead of running the check.
    $post_id = wp_insert_post(
      array(
        'post_title'   => 'Test post',
        'post_content' => 'This is a test post.',
        'post_type'    => 'post',
      )
    );

    // Return the clean up function.
    return function() use ( $post_id ) {
      // Remove the test post created.
      wp_delete_post( $post_id );
    }
  }
}
```

### Preparation Classes

Preparation classes are individual classes created to handle specific preparation logic in a single place.

> The [`Force_Single_Plugin_Preparation`](https://github.com/10up/plugin-check/blob/trunk/includes/Checker/Preparations/Force_Single_Plugin_Preparation.php) in the Plugin Checker is an example of a preparation class.

Reviewer: Better to link to the `Demo_Posts_Creation_Preparation` as that is a preparation that is more realistically encouraged to use. No check should really use the `Force_Single_Plugin_Preparation`, since that only makes sense at a global level.

Author: Updated

Preparation classes are useful for implementing the same preparation logic across multiple checks using the concept of shared preparations.

> ### Shared Preparations

Reviewer: See above, these two sections should be combined into a single section "Shared preparations". Essentially, you have to write a preparation class, and then any check can (re-)use it, so that's why it is called a shared preparation. And of course there are already preparation classes in the Plugin Checker itself, which can be used as a shared preparation.

Also see https://github.com/10up/plugin-check/pull/148#discussion_r1204839818 where I mentioned this earlier.

Author: Thanks @felixarntz I've combined these section and added additional detail

Shared preparations are used to prevent running the same preparations multiple times.

Before running checks against a plugin all shared preparations are collected and processed to remove any shared preparations that are the same. 

Check classes can define the shared preparations the use by using the `With_Shared_Preparation` interface.

The Check class should then implement the get_shared_preparations() method defined by the interface. This method returns an map of shared preparations where the preparation class name is the key and an array of constructor parameters as the value.

Below is an example of how the `Enqueued_Scripts_Size_Check` uses shared preparations.

```php

>   /**
>    * Returns an array of shared preparations for the check.
>    *
>    * @return array Returns a map of $class_name => $constructor_args pairs. If the class does not
>    *               need any constructor arguments, it would just be an empty array.
>    */
>   public function get_shared_preparations() {
>     $demo_posts = array_map(
>       function( $post_type ) {
>         return array(
>           'post_title'   => "Demo {$post_type} post",
>           'post_content' => 'Test content',
>           'post_type'    => $post_type,
>           'post_status'  => 'publish',
>         );
>       },
>       $this->viewable_post_types
>     );
> 
>     return array(
>       Demo_Posts_Creation_Preparation::class => array( $demo_posts ),
>     );
>   }

Reviewer 3: ```suggestion
/**
 * Returns an array of shared preparations for the check.
 *
 * @return array Returns a map of $class_name => $constructor_args pairs. If the class does not
 *               need any constructor arguments, it would just be an empty array.
 */
public function get_shared_preparations() {
  $demo_posts = array_map(
    function( $post_type ) {
      return array(
        'post_title'   => "Demo {$post_type} post",
        'post_content' => 'Test content',
        'post_type'    => $post_type,
        'post_status'  => 'publish',
      );
    },
    $this->viewable_post_types
  );

  return array(
    Demo_Posts_Creation_Preparation::class => array( $demo_posts ),
  );
}
```

```

> 

Reviewer: We need to add another section here for "Global preparations". This only needs to be an explanation that the Plugin Checker also includes some preparation classes that are used to prepare the overall environment, i.e. before _any_ runtime check. Global preparations cannot be controlled outside of the Plugin Checker at this point. An example for a global preparation is the `Force_Single_Plugin_Preparation` class, which ensures that only the plugin to check is actually active.

## Add the Check to the Plugin Checker

In order to run the check as part of the Plugin Checker process it needs to be added to the Plugin Checkers list of available checks.

This is done by using the `wp_plugin_check_checks` filter to register an instance of the check with its slug.

- If you're contributing to the actual plugin checker, add the check to the list in the `Abstract_Check_Runner::register_checks()` method.
- If you're implementing a check in code outside of the actual plugin checker, use the approach you've described here so far.

```php
add_filter(
  'wp_plugin_check_checks',
  function ( array $checks ) {
    // Add the check to the map of all available checks.
    $checks[ 'runtime_check' ] = new Runtime_Check();

    return $checks;
  }
)

> ```

Reviewer: Please remove all this content and instead link to the other article. No need for all of this to be duplicated. You can simply say something like "Adding a runtime check works similarly to adding a static check.", and link to the other article.

Author: Thanks @felixarntz I've removed and linked to the relevant section

  - `$line (int)` - The line on which the message occurred. Default 0 (unknown line).
  - `$column (int)` - The column on which the message occurred. Default 0 (unknown column).

In addition to adding messages, the `Check_Result` instance also provides access to the `Plugin_Context`. The plugin context is useful for getting the plugins path and urls when performing checks.

Below is an example showing how to access the plugin context and add messages using the `Check_Results` instance.

```php
/**
 * Runs the check on the plugin and amends results.
 *
 * @since n.e.x.t
 *
 * @param Check_Result $result The check results to amend and the plugin context.
 */
public function run( Check_Result $result ) {

  // Get the absolute file path for a specific file in the plugin.
  $plugin_file = $result->plugin()->path( 'plugin-file.php' );

  // Run logic to test the plugin for warnings/errors...

  // When an issue is found add a warning.
  $result->add_message(
    false, 
    'Warning message content.', 
    array(
      'code'   => 'warning_code',
      'file'   => $pluging_file,
      'line'   => 1,
      'column' => 10,
    )
  );
}

> ```

Reviewer: See above, let's remove the duplicate content and instead link to the other article. Also see https://github.com/10up/plugin-check/pull/148#discussion_r1204841140

Author: Thanks @felixarntz I've removed and linked to the relevant section
